Cython/Compiler/Parsing.py

```diff
@@ -2144,6 +2144,52 @@ def p_with_statement(s):
 
 
 def p_with_items(s, is_async=False):
+    """
+    Copied from CPython:
+    | 'with' '(' a[asdl_withitem_seq*]=','.with_item+ ','? ')' ':' b=block {
+        _PyAST_With(a, b, NULL, EXTRA) }
+    | 'with' a[asdl_withitem_seq*]=','.with_item+ ':' tc=[TYPE_COMMENT] b=block {
+        _PyAST_With(a, b, NEW_TYPE_COMMENT(p, tc), EXTRA) }
+    Therefore the first thing to try is the bracket-enclosed
+    version and if that fails try the regular version
+    """
+    brackets_succeeded=False
+    if s.sy == '(':
+        items = []
+        with tentatively_scan(s) as errors:
+            s.next()
+            while True:
+                items.append(p_with_item(s, is_async))
+                if s.sy == ")":
+                    s.next()
+                    break
+                s.expect(",")
+                if s.sy == ")":
+                    # trailing commas allowed
+                    s.next()
+                    break
+        brackets_succeeded = not errors
+    if not brackets_succeeded:
```

Comment:
Do we really need to parse twice here? Can't we just expect a closing parenthesis if there was an opening one, and otherwise just collect comma separated expressions?

Reply:
I think the difficulty is (and why it didn't work in the LL parser)

```
with (a+b), c:
```

It becomes really difficult to tell if the brackets are part of the first expression, or if they're around the whole thing.

Reply:
Although – hang on, that's what we did before (without the parentheses). What was wrong with that again?

Reply:
It doesn't support

```
with (cm1 as a, cm2):
```

So it'd assume that any bracket had to be enclosing an expression, and fail because the brackets were still open when it got to `as`.

Also
```
with (a, b, c):
```

Is pretty ambiguous... It's either 3 context managers or a single with statement with a tuple as a context manager. In the new grammar it ends up as 3 context managers because that's what succeeds first, but you don't know which one it'll be as you parse it

Reply:
> I think the difficulty is (and why it didn't work in the LL parser)
> 
> ```
> with (a+b), c:
> ```
> 
> It becomes really difficult to tell if the brackets are part of the first expression, or if they're around the whole thing.

Yes, it's not a simple LL thing. However, we have a hand-written parser, not a generated one, so isn't that just
1) find parenthesis or not, set a flag
2) parse CM sequence (at least one)
3)
    a) if closing parenthesis is not found and flag is not set: parse suite
    b) if closing parenthesis found != flag: error
    c) if closing parenthesis is found and flag is set:
        * if there was no comma yet, discard the parentheses and allow more CMs afterwards (you can't have `(a,b,c)` or `(a,)` as valid CM expressions)
        * if there was a comma, require suite

Reply:
An opening bracket doesn't necessarily enclose the first CM expression: `with (a+b).func(), c:`

So in stage 2 of your scheme it isn't clear if you need to include the bracket when parsing the first CM expression. In this case you must include the opening bracket or the parsing fails. But in the case of `with (a as x):` it'd be wrong to parse the bracket as part of the first CM expression.

> (you can't have (a,b,c) or (a,) as valid CM expressions)

They're grammatically OK. They'll fail at runtime because a tuple doesn't have `__enter__` and `__exit__`.

------------------------------

The main thing to be said for the way I've done it is that it is how the current version of the Python grammar works... it tries to parse the bracket-enclosed version and if that fails tries to parse the non-bracket enclosed version. So although parsing twice is inefficient when you have to do it, it's fairly easily linked back to the Python grammar.

Reply:
Ok. I understand that we'd probably always find enough examples to break any other scheme than trial and error.